### PR TITLE
[ROCm][Perf] DSv3.2 MI355X TP4 decode-step orchestration cleanup (3 micro-opts)

### PR DESCRIPTION
**Pull request by frida-andersson**
## TL;DR

Three independent CPU-side micro-optimizations on the DeepSeek-V3.2 AITER decode hot path. All are dispatch-side changes — no GPU kernel modifications.

| Workload | Pre-patch | This PR | Δ |
|---|---:|---:|---:|
| 5k/500/64 (956-step Kineto aggregate) | 41.522 ms/step | 40.234 ms/step | −1.288 ms (−3.10%) |
| GSM8K full set, flexible-extract (n=5) | — | **0.9378 ± 0.0067** | — (consistent with baseline) |

### (1) Router bf16 dispatch — `vllm/model_executor/models/deepseek_v2.py`

Eliminate `bfloat16tofloat32_copy_kernel_cuda` (58 launches/step on DSv3.2 TP4) on the ROCm AITER MoE routing path. The cast exists only because vLLM was asking the gate to emit fp32 router logits, while AITER's `biased_grouped_topk` kernel dispatches over `VLLM_DISPATCH_FLOATING_TYPES` and accepts bf16 directly (it promotes to fp32 internally during accumulation).

```python
# Before
self.gate.set_out_dtype(torch.float32)
# After
self.gate.set_out_dtype(self.gate.weight.dtype)   # bf16
```

The existing post-init cast at lines 358–362 then casts `e_score_correction_bias` from fp32 to bf16 at startup, satisfying the AITER kernel's same-`scalar_t` contract.

### (2) Sparse-MLA metadata cache — `vllm/v1/attention/backends/mla/rocm_aiter_mla_sparse.py`

`aiter::get_mla_metadata_v1` is launched once per `AttentionMetadataBuilder.build()` and computes a work-distribution schedule that is a deterministic function of `(num_tokens, max_query_len, num_heads, min(seq_lens, topk_tokens))`. Fingerprint those inputs CPU-side and skip the launch when the schedule would be identical. For long-context decode (`seq_lens ≥ topk_tokens=2048`) the fingerprint is shape-determined; cache hit rate is ~100% across the steady-state window.

### (3) Sparse-MLA tail-fill — same file

Three full-buffer `fill_(0)` launches on persistent attention-metadata buffers (`req_id_per_token_buffer`, `paged_kv_indices`, `paged_kv_indptr`) are replaced with shrink-tail-only fills. The redundant `paged_kv_indptr.fill_(0)` is dropped entirely — the `cumsum + scalar_broadcast` two lines below fully rewrites the range.

## Trace verification

Both before/after pairs use the same Perfetto search across the same captured 956-step decode window. The Omnibox match counter at the bottom does the heavy lifting.

### A. Metadata cache — `kn_get_mla_metadata_v1_2` call count

**Pre-patch:** Omnibox finds **1023** matches of `kn_get_mla_metadata_v1_2`. A metadata kernel slice is visible inside the highlighted `execute_context_0(0)_generation_64(64)` decode step on stream 3.
Screenshot 2026-05-18 at 15 04 01

**Post-patch:** same search finds only **64** matches — a 94% reduction. The 64 surviving calls are the prefill-phase invocations (one per incoming request from `max_concurrency=64`); during steady-state decode the cache hit rate is ~100%. No metadata kernel inside the highlighted decode step.
Screenshot 2026-05-18 at 15 04 49

### B. Router bf16 dispatch — `bfloat16tofloat32_copy_kernel_cuda` removal

**Pre-patch:** Omnibox finds **60,343** matches of `bfloat16tofloat32_copy_kernel_cuda` (≈ 58 MoE layers × 956 decode steps + prefill calls). The selected slice (green border) sits *between* the bf16 gate GEMM `wvSplitKrc<__hip_bfloat16, …>` and the router sort `aiter::grouped_topk_opt_sort_kernel<float, …>`. Current Selection panel confirms the kernel identity is the bf16→fp32 cast.
Screenshot 2026-05-18 at 15 06 36

**Post-patch:** the kernel chain on the same decode step is now `wvSplitKrc<__hip_bfloat16, …>` → `aiter::grouped_topk_opt_sort_kernel<c10::BFloat16, …>` directly, with no cast slice in between. The cast kernel is completely removed from the decode hot path.
Screenshot 2026-05-18 at 15 07 17

### ISL/OSL/MC 5k/500/64

Apples-to-apples comparison against the same workload run pre-patch:

| Metric | Pre-patch | This PR | Δ |
|---|---:|---:|---:|
| Decode step time (avg) | 41.522 ms | **40.234 ms** | **−1.288 ms (−3.10%)** |
| Total decode time (956 steps) | 39.695 s | 38.463 s | −1.232 s |
| `aiter::get_mla_metadata_v1` launches | 1023 | **64** | **−94%** |
| `bfloat16tofloat32_copy_kernel_cuda` launches | 60,343 | **0** | eliminated |

### Accuracy validation — GSM8K (full set, 1319 examples)

`lm_eval --model local-chat-completions --model_args model=deepseek-ai/DeepSeek-V3.2,base_url=http://127.0.0.1:8000/v1/chat/completions,tokenized_requests=False,num_concurrent=1000 --tasks gsm8k --apply_chat_template`. n-shot=5, fewshot-as-multiturn, temperature=0, batch_size=1.

| Filter | exact_match | Stderr |
|---|---:|---:|
| `flexible-extract` | **0.9378** | ± 0.0067 |
| `strict-match` | **0.9136** | ± 0.0077 |

Consistent with the published DeepSeek-V3.2 GSM8K baseline. **No accuracy regression** within stderr.

## Test plan

- [x] DSv3.2 TP4 decode TPOT improvement on 5k/500/64 (956-step aggregate: **−1.288 ms/step, −3.10%**).
- [x] DSv3.2 TP4 decode TPOT improvement on 1k/100/4 (776-step aggregate: **−78 µs/step**).
- [x] **Trace verification:** `aiter::get_mla_metadata_v1` call count drops from 1023 to 64 on the 956-step 5k aggregate (see screenshot pair A).
- [x] **Trace verification:** `bfloat16tofloat32_copy_kernel_cuda` is absent from the decode hot path (see screenshot pair B).
- [x] **Trace verification:** `grouped_topk_opt_sort_kernel` is instantiated on `c10::BFloat16` (not `float`) — visible in pair B's post-patch image.
- [x] **Trace verification:** `FillFunctor<int>` count drops by exactly 3/step on 1k single-step trace.
- [x] **Accuracy validation:** GSM8K full set — `flexible-extract 0.9378 ± 0.0067 / strict-match 0.9136 ± 0.0077`.

### CODE REVIEW

File: vllm/v1/attention/backends/mla/rocm_aiter_mla_sparse.py [hunk 1]
```diff
@@ -470,11 +481,32 @@ def build(
...
+        # defined by `paged_kv_indptr` (cumsum-built below), and the indexer
+        # kernel writes exactly those ranges before the attention kernel
+        # consumes them, so stale entries beyond the active range from a
+        # previous step are never observed.
+        new_req_extent = int(req_id_per_token.shape[0])
+        new_indices_extent = num_tokens * self.topk_tokens
+        if self._prev_req_extent > new_req_extent:
+            self.req_id_per_token_buffer[
+                new_req_extent : self._prev_req_extent
+            ].fill_(0)
```
[the remark this answers is not in the document]
Comment (frida-andersson) [reply vllm/v1/attention/backends/mla/rocm_aiter_mla_sparse.py]: Fixed in the follow-up commit ([a5af920](https://github.com/vllm-project/vllm/pull/42982/commits/a5af920)), but with torch.empty → torch.zeros at the allocator instead of adding a first-call branch in build(). Equivalent safety guarantee (the buffer is zero everywhere the kernel could observe), smaller diff, no branch in the hot path, and it mirrors the sister buffer paged_kv_indices which already uses torch.zeros (line 392).

File: vllm/v1/attention/backends/mla/rocm_aiter_mla_sparse.py [hunk 2]
```diff
@@ -505,27 +541,63 @@ def build(
...
+        # schedule would be identical to the previous step.
+        #
+        # `common_attn_metadata.seq_lens_cpu` is a property that lazily
+        # materializes from `self.seq_lens.to("cpu")` and is guaranteed
+        # non-None on access (see CommonAttentionMetadata in
+        # vllm/v1/attention/backend.py). It is therefore safe to index
+        # directly without a None branch -- a None branch would produce a
+        # cache key that ignores actual schedule changes (stale-metadata
+        # correctness bug) and would only be reachable if that property
+        # contract were later broken.
```
Comment (AndreasKaratzas) [inline vllm/v1/attention/backends/mla/rocm_aiter_mla_sparse.py]: Why are there so many comments? Like for example the entire last paragraph, I don't even understand what it's trying to convey. Please minimize AI generated comments across this entire PR. If a comment is not useful for a human, then delete it.

**Review by AndreasKaratzas — CHANGES_REQUESTED** [review]

File: vllm/v1/attention/backends/mla/rocm_aiter_mla_sparse.py (hunk 2 above)
Comment (frida-andersson) [reply vllm/v1/attention/backends/mla/rocm_aiter_mla_sparse.py]: Fair point! Pushed `fadadfc` which cuts ~35 lines of comments

File: vllm/v1/attention/backends/mla/rocm_aiter_mla_sparse.py [hunk 3]
```diff
@@ -375,7 +375,10 @@ def __init__(
             (1, 1), dtype=torch.int32, device=self.device
         )
 
-        self.req_id_per_token_buffer = torch.empty(
+        # zeros (not empty): build() only re-zeros the shrink-tail, and
+        # CUDA-graph-captured kernels may read padded positions beyond the
+        # active token range.
```
Comment (AndreasKaratzas) [inline vllm/v1/attention/backends/mla/rocm_aiter_mla_sparse.py]: This can be deleted too. It's well known that for ROCm preferred initialization is zeros.

File: vllm/v1/attention/backends/mla/rocm_aiter_mla_sparse.py [hunk 4]
```diff
@@ -458,6 +461,12 @@ def __init__(
             device=device,
         )
 
+        # Per-step memo state for the shrink-tail fills and the
+        # get_mla_metadata_v1 cache in build().
```
Comment (AndreasKaratzas) [inline vllm/v1/attention/backends/mla/rocm_aiter_mla_sparse.py]: This is also unnecessary as the vars have good self explanatory names.

File: vllm/v1/attention/backends/mla/rocm_aiter_mla_sparse.py [hunk 5]
```diff
@@ -505,27 +526,44 @@ def build(
         # treated as its own batch entry), so persistent metadata can always
         # be precomputed here. The kernel switches to the persistent
         # work-stealing path automatically when work_meta_data is non-None.
-        from aiter import get_mla_metadata_v1
-
-        get_mla_metadata_v1(
-            qo_indptr,
-            paged_kv_indptr,
-            paged_kv_last_page_len,
+        #
```
Comment (AndreasKaratzas) [inline vllm/v1/attention/backends/mla/rocm_aiter_mla_sparse.py]: nit: no reason for empty comment line

File: vllm/model_executor/models/deepseek_v2.py [hunk 6]
```diff
@@ -304,10 +304,11 @@ def __init__(
...
             and self.gate.e_score_correction_bias is not None
         ):
-            # AITER biased_grouped_topk requires the correction bias dtype to
-            # match the router logits. Keep DeepSeek's correction bias in fp32
-            # by requesting fp32 router logits for this routing path.
-            self.gate.set_out_dtype(torch.float32)
+            # Emit bf16 router logits to match the gate's native dtype. AITER's
+            # biased_grouped_topk dispatches on the gating_output dtype and
+            # requires the correction bias to match; this avoids a per-layer
+            # bf16->fp32 cast (the kernel still accumulates in fp32).
```
Comment (AndreasKaratzas) [inline vllm/model_executor/models/deepseek_v2.py]: Transpose to something sorter like "Accumulates in fp32; avoids bf16->fp32 cast"

**Review by AndreasKaratzas — CHANGES_REQUESTED** [review]

**Review by AndreasKaratzas — APPROVED** [review]
LGTM

**Comment by tjtanaa** [comment]
I will come back to this once we have the accuracy bugfix for DSA models in this PR https://github.com/vllm-project/vllm/pull/43781 .

After this PR, please rerun the accuracy test using num_shot 20, full GSM8k dataset. We should prioritize accuracy first over speed.

```
MODEL=deepseek-ai/DeepSeek-V3.2
lm_eval --model local-completions --model_args model=$MODEL,base_url=http://0.0.0.0:8894/v1/completions,num_concurrent=256,max_retries=10,max_gen_toks=2048,max_length=1048576,timeout=60000,trust_remote_code=True --batch_size auto --tasks gsm8k --num_fewshot 20 \
  --output_path ./results_dsv32_numshot20_cc256 \
  --log_samples \
| tee lmeval_dsv32_numshot20_cc256.log
```

**Comment by tjtanaa** [comment]
@frida-andersson please provide the accuracy score now that the accuracy bugfix has landed.

**Comment by frida-andersson** [comment]
@tjtanaa Re-running GSM8K with the full #42982 stack on top of the #43781 graft:

| Filter           | n-shot | Metric        | Value    | Stderr  |
|------------------|--------|---------------|----------|---------|
| flexible-extract | 20     | exact\_match   | 0.9522   | ±0.0059 |
| strict-match     | 20     | exact\_match   | 0.9530   | ±0.0058 |

**Review by tjtanaa — APPROVED** [review]
LGTM, @frida-andersson  Thanks for rerunning the tests. For DeepSeek V3.2 and GLM-5.1-FP8, moving forward, please use this lmeval command. High concurrency can pick up topk logic issue in the indexer. Large num shot (need to be at least 20) to trigger the sparse indexer.

```
MODEL=deepseek-ai/DeepSeek-V3.2
lm_eval --model local-completions --model_args model=$MODEL,base_url=http://0.0.0.0:8894/v1/completions,num_concurrent=256,max_retries=10,max_gen_toks=2048,max_length=1048576,timeout=60000,trust_remote_code=True --batch_size auto --tasks gsm8k --num_fewshot 20 \
  --output_path ./results_dsv32_numshot20_cc256 \
  --log_samples \
| tee lmeval_dsv32_numshot20_cc256.log
```
